src/helpers.c

```diff
@@ -149,130 +149,108 @@ void concat_str_array(char **output, char *const argv[])
+			assert(0);
 		}
+
+		int lhsresult = wait_process_terminate(lhspid);
+		int rhsresult = wait_process_terminate(rhspid);
+		exit(rhsresult != EXIT_SUCCESS ? rhsresult : lhsresult);
 	}
+	return wait_process_terminate(monitorpid);
+}
 
-	return status;
+int system_argv(char *const argv[])
+{
+	return system_argv_fd(-1, -1, -1, argv);
 }
 
-int system_argv_fd(char *const argv[], int newstdin, int newstdout, int newstderr)
+int system_argv_fd(int newstdinfd, int newstdoutfd, int newstderrfd, char *const cmd[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
-
-	pid = fork();
-
-	if (pid == 0) { /* child */
-		if (newstdin >= 0) {
-			if (dup2(newstdin, STDIN_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdin", "");
-				assert(0);
-			}
-			close(newstdin);
-		}
-		if (newstdout >= 0) {
-			if (dup2(newstdout, STDOUT_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdout", "");
-				assert(0);
-			}
-			close(newstdout);
-		}
-		if (newstderr >= 0) {
-			if (dup2(newstderr, STDERR_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stderr", "");
-				assert(0);
-			}
-			close(newstderr);
-		}
+	pid_t cmdpid = system_argv_fd_nowait(newstdinfd, newstdoutfd, newstderrfd, cmd);
+	return wait_process_terminate(cmdpid);
+}
 
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
+pid_t system_argv_fd_nowait(int newstdinfd, int newstdoutfd, int newstderrfd, char *const cmd[])
+{
+	pid_t cmdpid = fork();
+	if (cmdpid == -1) {
+		LOG(NULL, "Failed to fork to execute command", "%s", cmd[0]);
 		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
+	} else if (cmdpid == 0) {
+		set_std_fd(newstdinfd, newstdoutfd, newstderrfd);
+		execvp(*cmd, cmd);
+		LOG(NULL, "Command", "%s failed", cmd[0]);
 		assert(0);
-	} else {
-		pid_t ws = waitpid(pid, &child_exit_status, 0);
-
-		if (ws == -1) {
-			LOG(NULL, "Failed to wait for child process", "");
-			assert(0);
-		}
-
-		if (WIFEXITED(child_exit_status)) {
-			status = WEXITSTATUS(child_exit_status);
-		} else {
-			LOG(NULL, "Child process didn't exit", "");
-			assert(0);
-		}
-
-		if (status != 0) {
-			char *cmdline = NULL;
-
-			concat_str_array(&cmdline, argv);
-			LOG(NULL, "Failed to run command:", "%s", cmdline);
-			free(cmdline);
-		}
 	}
+	return cmdpid;
+}
 
-	return status;
+void set_std_fd(int newstdinfd, int newstdoutfd, int newstderrfd)
+{
+	if (newstdinfd != -1) {
+		move_fd(newstdinfd, STDIN_FILENO);
+	}
+	if (newstdoutfd != -1) {
+		move_fd(newstdoutfd, STDOUT_FILENO);
+	}
+	if (newstderrfd != -1) {
+		move_fd(newstderrfd, STDERR_FILENO);
+	}
 }
 
-int system_argv_pipe(char *const argvp1[], int stdinp1, int stderrp1,
-		     char *const argvp2[], int stdoutp2, int stderrp2)
+void move_fd(int oldfd, int newfd)
 {
-	int statusp2;
-	int pipefd[2];
+	if (dup2(oldfd, newfd) == -1) {
+		LOG(NULL, "Could not create duplicate file descriptor", "%d from %d", newfd, oldfd);
+		assert(0);
+	}
+	if (close(oldfd) == -1) {
```

Comment:
This is dangerous. Consider what happens if someone wants to say combine stdout and stderr and send them both down the pipe (i.e. "tar xvf fred |& tee /tmp/log"). Or what happens if oldfd and newfd are the same.

At the very least you need to just return doing nothing if they are the same.

Reply:
Oh, yes, I see.  The code assumes these to be different.  It doesn't handle the case where you really just want to symbolically link the file descriptors *and* keep them around.

Reply:
I would not log if they are the same, for 2 reasons. 1. The message is not actionable, you can't do anything if the message appears, and 2. it is not a bug. For example, if you ran 'make_fullfiles >&-' which closes stdout, assuming that the program made it as far as here then one of the file descriptors being returned by the pipe call could well be 1. Then you would want to move 1 to 1, which is a do nothing operation, not a bug.